Data/ByteString/Internal.hs

```diff
@@ -711,39 +726,59 @@ concat = \bss0 -> goLen0 bss0 bss0
    concat [x] = x
  #-}
 
--- | /O(log n)/ Repeats the given ByteString n times.
-times :: Integral a => a -> ByteString -> ByteString
-times n (BS fp len)
-  | n < 0 = error "stimes: non-negative multiplier expected"
+-- | Repeats the given ByteString n times.
+-- Polymorphic wrapper to make sure any generated
+-- specializations are reasonably small.
+stimesPolymorphic :: Integral a => a -> ByteString -> ByteString
+{-# INLINABLE stimesPolymorphic #-}
+stimesPolymorphic nRaw bs = case checkedToInt nRaw of
+  Just n  -> stimesInt n bs
+  _ | nRaw < 0
+      -- It may seem odd to check for negative input in both the
+      -- polymorphic wrapper and the specialized Int worker. But
+      -- checking here before converting to Int would not remove the
+      -- need to check inside the worker, since the Ord instance can
+      -- potentially behave arbitrarily. At the moment this isn't a
```

Comment:
This sentence is very unclear to me. Which `Ord` instance can potentially behave arbitrarily?

Also, if people write incorrect `Ord` instances, they can't expect `stimes` to work correctly. `bytestring` shouldn't waste code or long comments on this.

Reply:
There's exactly one `Ord` instance in play whose behavior we can't be sure of: The one coming from the `Integral` constraint on `nRaw`. But it's no trouble to be more explicit about this.

> Also, if people write incorrect `Ord` instances, they can't expect `stimes` to work correctly. `bytestring` shouldn't waste code or long comments on this.

Perhaps they can't expect `stimes` to work correctly, but they should be able expect `stimes` to at least be memory-safe. `Integral` is a lawless class. It's not clear to me if there's any expectation that the ordering on an `Integral` instance be compatible with the ordering suggested by its `toInteger` method. Is ordering the integers by some [p-adic valuation](https://en.wikipedia.org/wiki/P-adic_order) (for example) so exotic that it's "incorrect?"

I'd be happy to move this comment to a note somewhere and out of this function body, especially since there will be a similar need for a similar comment with the `stimes` implementation for `ShortByteString` when I get around to improving that more meaningfully.

Reply:
Thanks for the explanation! I doubt that there are many `Integral` instances out there where the `Ord` instance is incompatible with `toInteger`, but it seems possible in principle.

Pointing this out in the comment should be helpful.

If you move the note, please add a reference from `stimesInt`.